File: frontend/src/lib/api/dtos/event.ts
import { AttendanceActionType } from "@/lib/types/event/attendance";

interface Event {
  summary: string;
  location: string | null;
  dtstart: string;
  dtend: string;
  is_all_day: boolean;
  recurrence_list: string[];
  timezone: string;
}

interface EventWithId extends Event {
  id: string;
}

interface Attendance {
  action: AttendanceActionType;
  acted_at: string;
}

interface AttendancesWithUsername {
  username: string;
  attendances: Attendance[];
}

interface AttendanceTimeForecast {
  start: string;
  attended_at: string;
  duration: number;
}

export interface AttendanceTimeForecastsWithUsername {
  username: string;
  attendance_time_forecasts: AttendanceTimeForecast[];
}

export interface CreateEventRequest {
  event: Event;
}

export interface CreateEventResponse {
  error_codes: number[];
}

export interface UpdateEventRequest {
  event: Event;
}

export interface UpdateEventResponse {
  error_codes: number[];
}

export interface AttendEventRequest {
  action: AttendanceActionType;
}

export interface AttendEventResponse {
  error_codes: number[];
}

export interface UpdateAttendancesRequest {
  attendances: Attendance[];
}

export interface UpdateAttendancesResponse {
  error_codes: number[];
}

export interface GetAttendanceHistoryResponse {
  attendances_with_username: AttendancesWithUsername;
  error_codes: number[];
}

export interface GetMyEventsResponse {
  events: EventWithId[];
  error_codes: number[];
}

export interface GetFollowingEventsResponse {
  events: EventWithId[];
  error_codes: number[];
}

export interface GetGuestAttendanceStatusResponse {
  attend: boolean;
  error_codes: number[];
}

export interface GetAttendanceTimeForecastsResponse {
  attendance_time_forecasts_with_username: {
    [event_id: string]: {
      [user_id: number]: AttendanceTimeForecastsWithUsername;
    };
  };
  error_codes: number[];
}

export interface CreateOrUpdateGoalRequest {
  goal_text: string;
}

export interface CreateOrUpdateGoalResponse {
  error_codes: number[];
}

export interface GetGoalResponse {
  goal_text: string;
  error_codes: number[];
}

export interface CreateOrUpdateReviewRequest {
  review_text: string;
}

export interface CreateOrUpdateReviewResponse {
  error_codes: number[];
}

export interface GetReviewResponse {
  review_text: string;
  error_codes: number[];
}

File: backend/app/core/dtos/event.py
from datetime import datetime

from pydantic import BaseModel
from pydantic.fields import Field

from app.core.dtos.base import BaseModelWithErrorCodes
from app.core.features.event import AttendanceAction


class Event(BaseModel):
    summary: str = Field(..., title="Summary")
    location: str | None = Field(None, title="Location")
    dtstart: datetime = Field(..., title="DTSTART")
    dtend: datetime = Field(..., title="DTEND")
    is_all_day: bool = Field(True, title="Is All Day")
    recurrence_list: list[str] = Field(..., title="Recurrence List")
    timezone: str = Field(..., title="Timezone")


class EventWithId(Event):
    id: str = Field(..., title="Event ID")


class Attendance(BaseModel):
    action: AttendanceAction = Field(..., title="Attendance Action")
    acted_at: datetime = Field(..., title="Acted At")


class AttendancesWithUsername(BaseModel):
    username: str = Field(..., title="Username")
    attendances: list[Attendance] = Field(..., title="Attendances")


class AttendanceTimeForecast(BaseModel):
    start: datetime = Field(..., title="Event Started At")
    attended_at: datetime = Field(..., title="Attended At")
    duration: float = Field(..., title="Attendance Duration")


class AttendanceTimeForecastsWithUsername(BaseModel):
    username: str = Field(..., title="Username")
    attendance_time_forecasts: list[AttendanceTimeForecast] = Field(..., title="Attendance Time Forecasts")


class CreateEventRequest(BaseModel):
    event: Event = Field(..., title="Event")


class CreateEventResponse(BaseModelWithErrorCodes):
    pass


class UpdateEventRequest(BaseModel):
    event: Event = Field(..., title="Event")


class UpdateEventResponse(BaseModelWithErrorCodes):
    pass


class AttendEventRequest(BaseModel):
    action: AttendanceAction = Field(..., title="Attendance Action")


class AttendEventResponse(BaseModelWithErrorCodes):
    pass


class UpdateAttendancesRequest(BaseModel):
    attendances: list[Attendance] = Field(..., title="Attendances")


class UpdateAttendancesResponse(BaseModelWithErrorCodes):
    pass


class GetAttendanceHistoryResponse(BaseModelWithErrorCodes):
    attendances_with_username: AttendancesWithUsername = Field(..., title="Attendances with Username")


class GetMyEventsResponse(BaseModelWithErrorCodes):
    events: list[EventWithId] = Field(..., title="My Events")


class GetFollowingEventsResponse(BaseModelWithErrorCodes):
    events: list[EventWithId] = Field(..., title="Following Events")


class GetGuestAttendanceStatusResponse(BaseModelWithErrorCodes):
    attend: bool = Field(..., title="Is Attending")


class ForecastAttendanceTimeResponse(BaseModelWithErrorCodes):
    attendance_time_forecasts: dict[int, dict[str, list[AttendanceTimeForecast]]] = Field(
        ..., title="Attendance Time Forecasts"
    )


class GetAttendanceTimeForecastsResponse(BaseModelWithErrorCodes):
    attendance_time_forecasts_with_username: dict[str, dict[int, AttendanceTimeForecastsWithUsername]] = Field(
        ..., title="Attendance Time Forecasts with Username"
    )


class CreateOrUpdateGoalRequest(BaseModel):
    goal_text: str = Field(..., title="Goal Text")


class CreateOrUpdateGoalResponse(BaseModelWithErrorCodes):
    pass


class GetGuestGoalResponse(BaseModelWithErrorCodes):
    goal_text: str = Field(..., title="Goal Text")


class CreateOrUpdateReviewRequest(BaseModel):
    review_text: str = Field(..., title="Review Text")


class CreateOrUpdateReviewResponse(BaseModelWithErrorCodes):
    pass


class GetGuestReviewResponse(BaseModelWithErrorCodes):
    review_text: str = Field(..., title="Review Text")
